File: code/Advanced_Analytics_and_Machine_Learning-Chapter_24_Advanced_Analytics_and_Machine_Learning.scala
// in Scala
import org.apache.spark.ml.linalg.Vectors
val denseVec = Vectors.dense(1.0, 2.0, 3.0)
val size = 3
val idx = Array(1,2) // locations of non-zero elements in vector
val values = Array(2.0,3.0)
val sparseVec = Vectors.sparse(size, idx, values)
sparseVec.toDense
denseVec.toSparse


// COMMAND ----------

// in Scala
var df = spark.read.json("/data/simple-ml")
df.orderBy("value2").show()


// COMMAND ----------

spark.read.format("libsvm").load(
  "/data/sample_libsvm_data.txt")


// COMMAND ----------

// in Scala
import org.apache.spark.ml.feature.RFormula
val supervised = new RFormula()
  .setFormula("lab ~ . + color:value1 + color:value2")


// COMMAND ----------

// in Scala
val fittedRF = supervised.fit(df)
val preparedDF = fittedRF.transform(df)
preparedDF.show()


// COMMAND ----------

// in Scala
val Array(train, test) = preparedDF.randomSplit(Array(0.7, 0.3))


// COMMAND ----------

// in Scala
import org.apache.spark.ml.classification.LogisticRegression
val lr = new LogisticRegression().setLabelCol("label").setFeaturesCol("features")


// COMMAND ----------

// in Scala
println(lr.explainParams())


// COMMAND ----------

// in Scala
val fittedLR = lr.fit(train)


// COMMAND ----------

fittedLR.transform(train).select("label", "prediction").show()


// COMMAND ----------

// in Scala
val Array(train, test) = df.randomSplit(Array(0.7, 0.3))


// COMMAND ----------

// in Scala
val rForm = new RFormula()
val lr = new LogisticRegression().setLabelCol("label").setFeaturesCol("features")


// COMMAND ----------

// in Scala
import org.apache.spark.ml.Pipeline
val stages = Array(rForm, lr)
val pipeline = new Pipeline().setStages(stages)


// COMMAND ----------

// in Scala
import org.apache.spark.ml.tuning.ParamGridBuilder
val params = new ParamGridBuilder()
  .addGrid(rForm.formula, Array(
    "lab ~ . + color:value1",
    "lab ~ . + color:value1 + color:value2"))
  .addGrid(lr.elasticNetParam, Array(0.0, 0.5, 1.0))
  .addGrid(lr.regParam, Array(0.1, 2.0))
  .build()


// COMMAND ----------

// in Scala
import org.apache.spark.ml.evaluation.BinaryClassificationEvaluator
val evaluator = new BinaryClassificationEvaluator()
  .setMetricName("areaUnderROC")
  .setRawPredictionCol("prediction")
  .setLabelCol("label")


// COMMAND ----------

// in Scala
import org.apache.spark.ml.tuning.TrainValidationSplit
val tvs = new TrainValidationSplit()
  .setTrainRatio(0.75) // also the default.
  .setEstimatorParamMaps(params)
  .setEstimator(pipeline)
  .setEvaluator(evaluator)


// COMMAND ----------

// in Scala
val tvsFitted = tvs.fit(train)


// COMMAND ----------

evaluator.evaluate(tvsFitted.transform(test)) // 0.9166666666666667


// COMMAND ----------

// in Scala
import org.apache.spark.ml.PipelineModel
import org.apache.spark.ml.classification.LogisticRegressionModel
val trainedPipeline = tvsFitted.bestModel.asInstanceOf[PipelineModel]
val TrainedLR = trainedPipeline.stages(1).asInstanceOf[LogisticRegressionModel]
val summaryLR = TrainedLR.summary
summaryLR.objectiveHistory // 0.6751425885789243, 0.5543659647777687, 0.473776...


// COMMAND ----------

tvsFitted.write.overwrite().save("/tmp/modelLocation")


// COMMAND ----------

// in Scala
import org.apache.spark.ml.tuning.TrainValidationSplitModel
val model = TrainValidationSplitModel.load("/tmp/modelLocation")
model.transform(test)


// COMMAND ----------

File: code/Java/Advanced_Analytics_and_Machine_Learning_Chapter_24_Advanced_Analytics_and_Machine_Learning.java
// This Java code was contributed by @neeleshkumar-mannur

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import org.apache.spark.ml.Pipeline;
import org.apache.spark.ml.PipelineModel;
import org.apache.spark.ml.PipelineStage;
import org.apache.spark.ml.classification.LogisticRegression;
import org.apache.spark.ml.classification.LogisticRegressionModel;
import org.apache.spark.ml.classification.LogisticRegressionTrainingSummary;
import org.apache.spark.ml.evaluation.BinaryClassificationEvaluator;
import org.apache.spark.ml.feature.RFormula;
import org.apache.spark.ml.feature.RFormulaModel;
import org.apache.spark.ml.linalg.Vector;
import org.apache.spark.ml.linalg.Vectors;
import org.apache.spark.ml.param.ParamMap;
import org.apache.spark.ml.tuning.ParamGridBuilder;
import org.apache.spark.ml.tuning.TrainValidationSplit;
import org.apache.spark.ml.tuning.TrainValidationSplitModel;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;

public class Advanced_Analytics_and_Machine_Learning_Chapter_24_Advanced_Analytics_and_Machine_Learning {

	public static void main(String[] args) throws IOException {
		SparkSession spark = SparkSession
				.builder()
				.master("local[*]")
				.appName("Chapter24AdvancedAnalyticsAndMachineLearning")
				.getOrCreate();

		// Creating a Dense Vector
		Vector denseVector = Vectors.dense(1.0, 2.0, 3.0);
		System.out.println("Dense Vector:" + denseVector.toString());

		// Creating a Sparse Vector
		int size = 3;
		int[] idx = new int[] { 1, 2 };
		double[] values = new double[] { 2.0, 3.0 };

		Vector sparseVector = Vectors.sparse(size, idx, values);
		System.out.println("Sparse Vector:" + sparseVector.toString());

		// Reading the simple json file into a Dataframe / Dataset
		Dataset<Row> df = spark.read().json("data/simple-ml");
		df.orderBy("value2").show();

		// Initializing RFormula
		RFormula supervised = new RFormula().setFormula("lab ~ . + color:value1 + color:value2");

		// Applying RFormula
		RFormulaModel fittedRF = supervised.fit(df);
		Dataset<Row> preparedDF = fittedRF.transform(df);
		preparedDF.show();

		// Splitting PreparedDF into Train DF and Test DF
		Dataset<Row>[] trainTestDataArray = preparedDF.randomSplit(new double[] { 0.7, 0.3 });
		Dataset<Row> train = trainTestDataArray[0];
		Dataset<Row> test = trainTestDataArray[1];

		// Displaying sample data from both train as well as test dataset
		train.show();
		test.show();
		
		// Applying Logistic Regression on the split data
		LogisticRegression logisticRegression = new LogisticRegression().setLabelCol("label")
				.setFeaturesCol("features");

		// Printing the params
		System.out.println(logisticRegression.explainParams());

		// Fitting Train dataset into the model
		LogisticRegressionModel fittedLR = logisticRegression.fit(train);
		fittedLR.transform(train).select("label","prediction").show();
		
		// Splitting the original dataset into train and test set
		Dataset<Row> [] trainTestSet = df.randomSplit(new double [] {0.7, 0.3});
		train = trainTestSet[0];
		test = trainTestSet[1];
		
		// Initializing RFormula
		RFormula rForm = new RFormula();

		// Reinitializing the Logistic Regression model since it is already trained on other data
		logisticRegression = new LogisticRegression().setLabelCol("label").setFeaturesCol("features");
		
		// Creating a Pipeline for execution
		PipelineStage[] stages = new PipelineStage[2];
		stages[0] = rForm;
		stages[1] = logisticRegression;

		Pipeline pipeline = new Pipeline().setStages(stages);

		// Applying Param Grid Builder
		List<String> paramValues = new ArrayList<String>();
		paramValues.add("lab ~ . + color:value1");
		paramValues.add("lab ~ . + color:value1 + color:value2");

		ParamMap [] params = new ParamGridBuilder()
				.addGrid(rForm.formula(), scala.collection.JavaConverters
						.asScalaIteratorConverter(paramValues.iterator()).asScala().toSeq())
				.addGrid(logisticRegression.elasticNetParam(), new double [] {0.0, 0.5, 1.0})
				.addGrid(logisticRegression.regParam(), new double [] {0.1, 2.0})
				.build();
		
		// Obtaining the Evaluator
		BinaryClassificationEvaluator evaluator = new BinaryClassificationEvaluator()
				.setMetricName("areaUnderROC")
				.setRawPredictionCol("prediction")
				.setLabelCol("label");
		
		// Creating Train Validation Split
		TrainValidationSplit trainValidationSplit = new TrainValidationSplit()
				.setTrainRatio(0.75)
				.setEstimatorParamMaps(params)
				.setEstimator(pipeline)
				.setEvaluator(evaluator);
		
		TrainValidationSplitModel tvsFitted = trainValidationSplit.fit(train);
		
		evaluator.evaluate(tvsFitted.transform(test));
		
		// Getting the best Model
		PipelineModel trainedPipeline = (PipelineModel) tvsFitted.bestModel();
		LogisticRegressionModel trainedLR = (LogisticRegressionModel) trainedPipeline.stages()[1];
		LogisticRegressionTrainingSummary summaryLR = trainedLR.summary();
		
		for(double objectiveHist: summaryLR.objectiveHistory()) {
			System.out.println(objectiveHist);
		}
		
		// Persisting the model to disk
		tvsFitted.write().overwrite().save("tmp/modelLocation");
		
		// Loading the persisted model from disk and testing
		TrainValidationSplitModel model = TrainValidationSplitModel.load("tmp/modelLocation");
		Dataset<Row> tested = model.transform(test);
		
		tested.show();
	}
}
